.github/workflows/pre-commit.yml

```diff
@@ -0,0 +1,29 @@
+# see https://github.com/pre-commit/action
+
+name: pre-commit
+
+on:
+  workflow_dispatch:
+  pull_request:
+  push:
+    branches:
+      - main
+
+jobs:
+  pre-commit:
+    name: Format
+    runs-on: ubuntu-22.04
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
+        with:
+          python-version: '3.10'
+      - name: Install clang-format-14
+        run: sudo apt-get install clang-format-14
+      - uses: pre-commit/action@v3.0.1
+        id: precommit
+      - name: Upload pre-commit changes
+        if: failure() && steps.precommit.outcome == 'failure'
```

Comment:
You can't push changes on a pull_request branch. Also we shouldn't let CI push on the main.